File: mobile/lib/core/utils/date.dart
import 'package:intl/intl.dart';

import 'timezone.dart';

/// Local calendar date as `YYYY-MM-DD` (device / screen timezone).
String todayYmdLocal([DateTime? dt]) => todayYmdInScreenZone(dt);

/// Add [days] to a `YYYY-MM-DD` string (returns empty if invalid).
String addDaysYmd(String ymd, int days) {
  final dt = parseApiDate(ymd);
  if (dt == null) return '';
  return todayYmdLocal(dt.add(Duration(days: days)));
}

/// Date parsing helpers for API payloads.
DateTime? parseApiDate(dynamic value) {
  if (value == null) return null;
  if (value is DateTime) return value;
  final s = value.toString().trim();
  if (s.isEmpty) return null;
  return DateTime.tryParse(s);
}

String? isoDateOnly(dynamic value) {
  final dt = parseApiDate(value);
  if (dt == null) return null;
  final y = dt.year.toString().padLeft(4, '0');
  final m = dt.month.toString().padLeft(2, '0');
  final d = dt.day.toString().padLeft(2, '0');
  return '$y-$m-$d';
}

String _monthYearLabel(DateTime d) => DateFormat('MMM yyyy').format(d);

/// Calendar days left in [from]'s month (including today).
int daysRemainingInMonth([DateTime? from]) {
  final n = from ?? DateTime.now();
  final lastDay = DateTime(n.year, n.month + 1, 0).day;
  return (lastDay - n.day).clamp(0, 31);
}

/// Display label for the active calendar month, e.g. "May 2026".
String currentMonthYearLabel([DateTime? from]) {
  final n = from ?? DateTime.now();
  return DateFormat('MMMM yyyy').format(n);
}

/// Column headers for dashboard sales-compare bars (left → right):
/// same month last year · previous calendar month · active period.
List<String> momCompareColumnLabels({
  required String dateToYmd,
  required String dateFromYmd,
  required String preset,
}) {
  final end = parseApiDate(dateToYmd) ?? DateTime.now();
  final start = parseApiDate(dateFromYmd) ?? end;

  final lastYearLabel = _monthYearLabel(DateTime(end.year - 1, end.month));
  final lastMonthLabel = _monthYearLabel(DateTime(end.year, end.month - 1));

  final String currentLabel;
  switch (preset) {
    case 'TODAY':
      currentLabel = DateFormat('d MMM yyyy').format(end);
    case 'WEEK':
      if (start.year == end.year && start.month == end.month) {
        currentLabel =
            '${start.day}–${end.day} ${DateFormat('MMM yyyy').format(end)}';
      } else {
        currentLabel =
            '${DateFormat('d MMM').format(start)} – ${DateFormat('d MMM yyyy').format(end)}';
      }
    case 'QUARTER':
      final q = ((end.month - 1) ~/ 3) + 1;
      currentLabel = 'Q$q ${end.year}';
    case 'YEAR':
      currentLabel = '${end.year}';
    default:
      if (dateFromYmd.isNotEmpty &&
          dateFromYmd != dateToYmd &&
          (start.year != end.year || start.month != end.month)) {
        currentLabel =
            '${DateFormat('d MMM').format(start)} – ${DateFormat('d MMM yyyy').format(end)}';
      } else {
        currentLabel = _monthYearLabel(end);
      }
  }

  return [lastYearLabel, lastMonthLabel, currentLabel];
}

File: frontend/src/utils/date.js
import { screenTimeZone, todayYmdInScreenZone, withScreenTimezone } from "./timezone.js";

export { screenTimeZone, todayYmdInScreenZone, withScreenTimezone };

function pad2(n) {
  return String(n).padStart(2, "0");
}

/** Local calendar date as YYYY-MM-DD in the screen timezone. */
export function todayYmdLocal(d = new Date()) {
  if (!d || d.getTime() === Date.now()) {
    return todayYmdInScreenZone();
  }
  const y = d.getFullYear();
  const m = pad2(d.getMonth() + 1);
  const day = pad2(d.getDate());
  return `${y}-${m}-${day}`;
}

export function isValidYmd(value) {
  const s = String(value || "").trim();
  if (!/^\d{4}-\d{2}-\d{2}$/.test(s)) return false;
  const [y, m, d] = s.split("-").map(Number);
  if (!Number.isFinite(y) || !Number.isFinite(m) || !Number.isFinite(d)) return false;
  const dt = new Date(y, m - 1, d);
  return dt.getFullYear() === y && dt.getMonth() === m - 1 && dt.getDate() === d;
}

/** Add calendar days to a YYYY-MM-DD string. Returns "" if input invalid. */
export function addDaysYmd(ymd, days) {
  const s = String(ymd || "").trim();
  if (!isValidYmd(s)) return "";
  const [y, m, d] = s.split("-").map(Number);
  const dt = new Date(y, m - 1, d + Number(days || 0));
  return todayYmdLocal(dt);
}

